### PR TITLE
Introducing engine/runtime-compiler project

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

`LocalScope` and `FramePointer` are the last classes from interpreter referenced by `org.enso.compiler` package. Splitting the `FrameDescriptor` building capability away from `LocalScope` and moving the remaining code to _IR compiler_ package. That allows us to create separate `runtime-compiler` project and completely isolate _IR compiler_ from the interpreter.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides
- All code has been tested:
  - [x] Unit tests continue to pass
  - [x] [Engine benchmarks](https://github.com/enso-org/enso/actions/runs/6716711770) and [2nd run](https://github.com/enso-org/enso/actions/runs/6717686731) show no regressions
  - [x] [StdLib benchmarks](https://github.com/enso-org/enso/actions/runs/6716719766) show no regressions

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1438,6 +1437,23 @@ lazy val `runtime-parser` =
...
+lazy val `runtime-compiler` =
+  (project in file("engine/runtime-compiler"))
+    .settings(
+      frgaalJavaCompilerSetting,
+      instrumentationSettings,
+      libraryDependencies ++= Seq(
+        "junit"          % "junit"           % junitVersion     % Test,
+        "com.github.sbt" % "junit-interface" % junitIfVersion   % Test,
+        "org.scalatest" %% "scalatest"       % scalatestVersion % Test,
+        "com.lihaoyi"   %% "fansi"           % fansiVersion
```
Comment (JaroslavTulach) [inline build.sbt:1449]: There are two remaining dependencies we want to get rid in `runtime-compiler` (if we want to run it in VSCode extension as #7054 suggests or in the browser as proposed by #8094):
- we need to remove dependency on `fansi`
- we need to remove dependency on Truffle's `SourceSection`

Let those refactorings be addressed separately by
- #8210

**Comment by JaroslavTulach** [comment]
Since now the _static (IR) compiler_ resides in `engine/runtime-compiler` project including all its import resolution, error detection, and optimization phases. The _dynamic compiler_ remains in `engine/runtime` (together with a lot of other stuff). CCing @radeusgd.

**Comment by JaroslavTulach** [comment]
Engine benchmarks seem to be fine except

AtomBenchmarks_benchReverseList

we may need [yet another run](https://github.com/enso-org/enso/actions/runs/6717686731) to be ensured everything is OK. The second run shows, everything is OK:

After 2nd run of AtomBenchmarks_benchReverseList

File: engine/runtime/src/main/java/org/enso/interpreter/node/EnsoRootNode.java [hunk 2]
```diff
@@ -49,6 +56,29 @@ protected EnsoRootNode(
...
+          descriptorBuilder.addSlot(
+            FrameSlotKind.Illegal,
+            definition.symbol(),
+            null
+            );
+            // assert(localFrameSlotIdxs(definition.id) == returnedFrameIdx)
+      }
+      descriptorBuilder.defaultValue(DataflowError.UNINITIALIZED);
+      var frameDescriptor = descriptorBuilder.build();
+      // assert(LocalScope.internalSlots.length + localFrameSlotIdxs.size == frameDescriptor.getNumberOfSlots)
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/node/EnsoRootNode.java]: I would suggest to either remove those commented asserts completely or (preferably) bring them in, as they were in the old code

**Review by hubertp — APPROVED** [review]
Minor nit

File: engine/runtime/src/main/java/org/enso/interpreter/node/EnsoRootNode.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/EnsoRootNode.java]: Removed in fd3a2286a1

**Pull request merged**